<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Intro to R - Cluster analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./correlation_linearregression.html" rel="next">
<link href="./customize-plots.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./clustering.html">7. Data analysis</a></li><li class="breadcrumb-item"><a href="./clustering.html">Cluster analysis</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Intro to R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">1. To do ahead of time</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installing R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./code-of-conduct.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Code of conduct</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./benchmarks.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Benchmarks</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">2. Introduction</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./file-location.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">File structure</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">What is ‘R’?</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rstudio-orientation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">RStudio</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">3. Coding in R</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basic-syntax.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic syntax</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-types.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data classes &amp; data frames</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./packages-libraries.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R packages</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">4. Manipulations of data</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./importing-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data import</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./exporting-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Export data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyverse-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tidyverse</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">5. Data wrangling</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./StarWars.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Star Wars data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./growthcurve.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Growth curve</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./asvdata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ASV data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">6. Data visualization</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basic-plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Basic plots</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./customize-plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Customize plots</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
 <span class="menu-text">7. Data analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./clustering.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Cluster analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./correlation_linearregression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Correlation &amp; Linear regression</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#why" id="toc-why" class="nav-link active" data-scroll-target="#why">Why?</a>
  <ul class="collapse">
  <li><a href="#ordinations" id="toc-ordinations" class="nav-link" data-scroll-target="#ordinations">Ordinations</a></li>
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  </ul></li>
  <li><a href="#import-data" id="toc-import-data" class="nav-link" data-scroll-target="#import-data">Import data</a>
  <ul class="collapse">
  <li><a href="#set-up-preliminary-environment" id="toc-set-up-preliminary-environment" class="nav-link" data-scroll-target="#set-up-preliminary-environment">Set up preliminary environment</a></li>
  </ul></li>
  <li><a href="#evidence-of-compositionality-diagnostic" id="toc-evidence-of-compositionality-diagnostic" class="nav-link" data-scroll-target="#evidence-of-compositionality-diagnostic">Evidence of compositionality (diagnostic)</a></li>
  <li><a href="#data-transformations" id="toc-data-transformations" class="nav-link" data-scroll-target="#data-transformations">Data transformations</a>
  <ul class="collapse">
  <li><a href="#isometric" id="toc-isometric" class="nav-link" data-scroll-target="#isometric">Isometric</a></li>
  <li><a href="#center-log-ratio" id="toc-center-log-ratio" class="nav-link" data-scroll-target="#center-log-ratio">Center log ratio</a></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  </ul></li>
  <li><a href="#pcoa" id="toc-pcoa" class="nav-link" data-scroll-target="#pcoa">PCoA</a>
  <ul class="collapse">
  <li><a href="#distance-matrix" id="toc-distance-matrix" class="nav-link" data-scroll-target="#distance-matrix">Distance matrix</a></li>
  <li><a href="#euclidean" id="toc-euclidean" class="nav-link" data-scroll-target="#euclidean">Euclidean</a></li>
  <li><a href="#nmds" id="toc-nmds" class="nav-link" data-scroll-target="#nmds">NMDS</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Cluster analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="why" class="level1">
<h1>Why?</h1>
<ol type="1">
<li><p>Too many dimensions</p></li>
<li><p>Need to sort and group</p></li>
<li><p>Test a hypothesis | Explore hypotheses</p></li>
</ol>
<section id="ordinations" class="level2">
<h2 class="anchored" data-anchor-id="ordinations">Ordinations</h2>
<p><em>Ordinations</em> - taking complex data and working to present it in 2-3 dimensions, we are placing a new coordinate system (fitting) in place. Principle Component Analysis (PCA) works to generate this coordinate system that is most appropriate for your data. Evaluated based on the eigen decomposition from sample covariance matrix. (Eigenvectors = principle axes)</p>
</section>
<section id="background" class="level2">
<h2 class="anchored" data-anchor-id="background">Background</h2>
<p><strong>References / recommended reading:</strong></p>
<ol type="1">
<li><p>THIS TUTORIAL: Coenen AR, Hu SK, Luo E, Muratore D, Weitz JS. A Primer for Microbiome Time-Series Analysis. Front Genet. 2020 Apr 21;11:310.</p></li>
<li><p>Gloor, G. B., Macklaim, J. M., Pawlowsky-Glahn, V. &amp; Egozcue, J. J. Microbiome Datasets Are Compositional: And This Is Not Optional. Front. Microbiol. 8, 57–6 (2017).</p></li>
<li><p>Weiss, S. et al.&nbsp;Normalization and microbial differential abundance strategies depend upon data characteristics. Microbiome 5, 1–18 (2017).</p></li>
<li><p>McMurdie, P. J. &amp; Holmes, S. Waste Not, Want Not: Why Rarefying Microbiome Data Is Inadmissible. PLoS Comput Biol 10, e1003531 (2014).</p></li>
<li><p>Silverman JD, Washburne AD, Mukherjee S, David LA. A phylogenetic transform enhances analysis of compositional microbiota data. eLife. 2017 Feb 15;6:e21887.</p></li>
</ol>
</section>
</section>
<section id="import-data" class="level1">
<h1>Import data</h1>
<section id="set-up-preliminary-environment" class="level2">
<h2 class="anchored" data-anchor-id="set-up-preliminary-environment">Set up preliminary environment</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'tidyr' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'readr' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'dplyr' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'stringr' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compositions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'compositions' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to compositions, a package for compositional data analysis.
Find an intro with "? compositions"


Attaching package: 'compositions'

The following objects are masked from 'package:stats':

    anova, cor, cov, dist, var

The following object is masked from 'package:graphics':

    segments

The following objects are masked from 'package:base':

    %*%, norm, scale, scale.default</code></pre>
</div>
</div>
</section>
</section>
<section id="evidence-of-compositionality-diagnostic" class="level1">
<h1>Evidence of compositionality (diagnostic)</h1>
<p>Do we need to transform our data? We can use the estimate covariance matrix</p>
<p>Prep data frame</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(asv_table)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># head(asv_table)</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Random subset 1000 rows</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>asv_df <span class="ot">&lt;-</span> asv_table <span class="sc">|&gt;</span> </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">1000</span>) <span class="sc">|&gt;</span> </span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"FeatureID"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>Taxon)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># dim(asv_df)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>%*%</code> = matrix multiplication sign in R; used here to multiply OTU/ASV data matrix to itself to estimate covariance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>covariance <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(asv_df) <span class="sc">%*%</span> <span class="fu">t</span>(asv_df)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>det_output <span class="ot">&lt;-</span> <span class="fu">det</span>(covariance)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>det_output</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<p>The determinant of the covariance matrix (what we just calculated) is equivalent to the product of the proportion of variance explained by every PCA axis. If the determinant is 0, that means there is an axis which explains 0 variance that we can’t separate from the other axes. Therefore, data need to be transformed to be suitable for PCA.</p>
</section>
<section id="data-transformations" class="level1">
<h1>Data transformations</h1>
<p>Another approach to PCA ordination w/Compositional Data: Log-Ratio Transformations</p>
<ul>
<li><p><strong>Log-ratio</strong> - combat issues with variance and distribution, all variables will be transformed to be relative to the abundance of an arbitrary focal taxon. Divide all the count data by the abundance of a focal taxa and take the natural log. So why do this? This is one way we can transform our amplicon data to be more appropriate/more suitable for many statistical approaches.</p></li>
<li><p><strong>ilr</strong> - isometric log ratio</p></li>
<li><p><strong>clr</strong> - centered log ratio, values will be centered around the geometric average</p></li>
</ul>
<section id="isometric" class="level2">
<h2 class="anchored" data-anchor-id="isometric">Isometric</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gorda_ilr <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(compositions<span class="sc">::</span><span class="fu">ilr</span>(<span class="fu">t</span>(asv_df)))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>cov_ilr <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(gorda_ilr) <span class="sc">%*%</span> <span class="fu">t</span>(gorda_ilr)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>det_ilr <span class="ot">&lt;-</span> <span class="fu">det</span>(cov_ilr)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>det_ilr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.426553e+40</code></pre>
</div>
</div>
<p><code>prcomp()</code> - Principle Component Analysis. Input is a data matrix and results are an R class “prcomp”.</p>
<p>Explore the covariance after doing the transformation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>ilr_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(gorda_ilr)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(ilr_pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "prcomp"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>ilr_pca<span class="sc">$</span>sdev </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 5.178359e+00 4.451509e+00 4.126228e+00 3.737649e+00 3.415436e+00
 [6] 3.367891e+00 3.237619e+00 3.204526e+00 3.121084e+00 3.036606e+00
[11] 2.996977e+00 2.899159e+00 2.780506e+00 2.618325e+00 2.560523e+00
[16] 2.146984e+00 1.862622e+00 2.388152e-15</code></pre>
</div>
</div>
<p><em>follow up reading on prcomp</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visual representation - screeplot</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>lograt_variances <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(ilr_pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span><span class="fu">sum</span>(ilr_pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">%&gt;%</span> <span class="co">#Extract axes</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Format to plot</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">PercVar =</span> <span class="st">'ilr_pca$sdev^2/sum(ilr_pca$sdev^2)'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"PCaxis"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  data.frame</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot screeplot</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(lograt_variances, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.numeric</span>(PCaxis), <span class="at">y =</span> PercVar)) <span class="sc">+</span> </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"grey"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>),</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"PC axis"</span>, <span class="at">y =</span> <span class="st">"% Variance"</span>, <span class="at">title =</span> <span class="st">"Log-Ratio PCA Screeplot"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="clustering_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="center-log-ratio" class="level2">
<h2 class="anchored" data-anchor-id="center-log-ratio">Center log ratio</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>gorda_clr <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(compositions<span class="sc">::</span><span class="fu">clr</span>(<span class="fu">t</span>(asv_df)))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>cov_clr <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(gorda_clr) <span class="sc">%*%</span> <span class="fu">t</span>(gorda_clr)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>det_clr <span class="ot">&lt;-</span> <span class="fu">det</span>(cov_clr)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>det_clr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2.426553e+40</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>clr_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(gorda_clr)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>clr_variances <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(clr_pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span><span class="fu">sum</span>(clr_pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">%&gt;%</span> <span class="co">#Extract axes</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Format to plot</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">PercVar =</span> <span class="st">'clr_pca$sdev^2/sum(clr_pca$sdev^2)'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"PCaxis"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  data.frame</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot screeplot</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clr_variances, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.numeric</span>(PCaxis), <span class="at">y =</span> PercVar)) <span class="sc">+</span> </span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"grey"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>),</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"PC axis"</span>, <span class="at">y =</span> <span class="st">"% Variance"</span>, <span class="at">title =</span> <span class="st">"Log-Ratio PCA Screeplot"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="clustering_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="pca" class="level2">
<h2 class="anchored" data-anchor-id="pca">PCA</h2>
<p>Lets pick one. My favorite is always the clr.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(clr_pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "prcomp"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the PCA</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># clr_pca$</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>clr_pca<span class="sc">$</span>x <span class="co">#View PC values</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                               PC1       PC2        PC3
GordaRidge_Vent039_SUPRS1_2019          -7.8136532  7.494693 -2.8582200
GordaRidge_Vent040_SUPRS2_2019          -9.1363140  4.538102  0.3794693
GordaRidge_BSW081_sterivex_2019          1.6009412  1.545660 -7.0441090
GordaRidge_BSW056_sterivex_2019_REPb     0.4809999 -5.890110 -0.8983877
GordaRidge_Vent088_SUPRS3_2019           5.2263501  1.191327  2.6590195
GordaRidge_Plume001_sterivex_2019_REPa   7.7946359  5.972814  6.6121580
GordaRidge_Plume001_sterivex_2019_REPb   5.0334438  4.528548  1.8486798
GordaRidge_Plume036_sterivex_2019_REPb   2.2261855  1.496157 -8.1632903
GordaRidge_Plume096_sterivex_2019        1.0336953 -6.241536 -1.4581866
GordaRidge_Vent105_SUPRS9_2019          -0.2470926 -3.422314 -5.4194807
GordaRidge_Vent041_SUPRS3_2019         -11.2811708  1.871824  4.4183622
GordaRidge_Vent086_SUPRS1_2019           4.5859168  1.494836 -1.1330836
GordaRidge_Vent107_SUPRS11_2019          2.1292156  0.118059  3.9059303
GordaRidge_Vent009_SUPRS1_2019          -0.3035928 -2.064704 -2.5717286
GordaRidge_Vent011_SUPRS3_2019          -4.1701019 -6.917408  5.0390290
GordaRidge_Vent106_SUPRS10_2019          2.6372677 -3.654706  2.6437269
GordaRidge_Vent010_SUPRS2_2019          -2.3235680 -5.455979  1.7994360
GordaRidge_Vent087_SUPRS2_2019           2.5268413  3.394736  0.2406755
                                              PC4        PC5        PC6
GordaRidge_Vent039_SUPRS1_2019          0.5853997 -1.8599013  0.8335125
GordaRidge_Vent040_SUPRS2_2019          0.9513623  0.9762615 -0.7453047
GordaRidge_BSW081_sterivex_2019        -0.5990361 -4.8957990  9.2979107
GordaRidge_BSW056_sterivex_2019_REPb   -3.8027815 -3.4603966 -3.9973122
GordaRidge_Vent088_SUPRS3_2019          3.1452520 -1.7983552 -0.9985789
GordaRidge_Plume001_sterivex_2019_REPa -5.9821424  0.5643154 -0.6252184
GordaRidge_Plume001_sterivex_2019_REPb -3.7657968 -0.2333092  0.6694650
GordaRidge_Plume036_sterivex_2019_REPb -1.1361132 10.5142376 -1.8330749
GordaRidge_Plume096_sterivex_2019       2.0373430 -0.2214074  0.5018464
GordaRidge_Vent105_SUPRS9_2019         -6.2391838 -3.7399956 -5.5263953
GordaRidge_Vent041_SUPRS3_2019         -0.7753195 -1.0841440 -2.3449245
GordaRidge_Vent086_SUPRS1_2019         10.0067007 -2.6197402 -4.9560969
GordaRidge_Vent107_SUPRS11_2019        -0.6898155  1.1138606  2.5978148
GordaRidge_Vent009_SUPRS1_2019         -0.1128133 -0.0803269  0.7626128
GordaRidge_Vent011_SUPRS3_2019          0.3078292  3.6034640  2.8101004
GordaRidge_Vent106_SUPRS10_2019         1.9887275 -0.5606825  1.6681576
GordaRidge_Vent010_SUPRS2_2019          0.6994025  2.0001555  2.2287919
GordaRidge_Vent087_SUPRS2_2019          3.3809852  1.7817633 -0.3433062
                                               PC7        PC8          PC9
GordaRidge_Vent039_SUPRS1_2019          0.10984998  0.3696854  0.715605912
GordaRidge_Vent040_SUPRS2_2019          0.59808106  1.1344267  2.859706232
GordaRidge_BSW081_sterivex_2019         0.08611798  1.1064133 -0.571670051
GordaRidge_BSW056_sterivex_2019_REPb    4.50619192  1.7292399  5.920244146
GordaRidge_Vent088_SUPRS3_2019          1.57717333 -6.7397281 -2.274740941
GordaRidge_Plume001_sterivex_2019_REPa -0.44204944  4.2474052 -2.591276630
GordaRidge_Plume001_sterivex_2019_REPb  0.17281861  2.5605504 -0.079034538
GordaRidge_Plume036_sterivex_2019_REPb  3.02235302 -0.4671150 -1.010926692
GordaRidge_Plume096_sterivex_2019       1.52238580  0.3511822 -6.053876020
GordaRidge_Vent105_SUPRS9_2019         -5.70963255 -2.7002895 -0.950053369
GordaRidge_Vent041_SUPRS3_2019          2.32677353 -1.7147593 -5.064775761
GordaRidge_Vent086_SUPRS1_2019          0.44824464  5.3025673  1.053929872
GordaRidge_Vent107_SUPRS11_2019         3.43525619 -5.3713132  5.899991073
GordaRidge_Vent009_SUPRS1_2019          0.22278117  0.7541454 -0.068182539
GordaRidge_Vent011_SUPRS3_2019         -4.98424817  3.3040684  1.407308877
GordaRidge_Vent106_SUPRS10_2019         1.62767517 -1.9783337 -1.492912037
GordaRidge_Vent010_SUPRS2_2019         -0.41941509  1.7196031  0.003326381
GordaRidge_Vent087_SUPRS2_2019         -8.10035714 -3.6077484  2.297336084
                                              PC10       PC11       PC12
GordaRidge_Vent039_SUPRS1_2019          1.39946675  3.2753120  1.8936698
GordaRidge_Vent040_SUPRS2_2019          0.14301966  3.0504403  4.4990535
GordaRidge_BSW081_sterivex_2019        -1.89264463 -2.7457225 -1.8060827
GordaRidge_BSW056_sterivex_2019_REPb   -3.55087935  3.4437896 -3.8670740
GordaRidge_Vent088_SUPRS3_2019         -7.12165231  0.5030758  3.9059055
GordaRidge_Plume001_sterivex_2019_REPa  0.15687793  0.3409309  0.5627682
GordaRidge_Plume001_sterivex_2019_REPb  0.16114175  0.5779229 -0.2249137
GordaRidge_Plume036_sterivex_2019_REPb -0.66714902 -1.3706984 -0.9992002
GordaRidge_Plume096_sterivex_2019       1.53890046  4.2112810  1.4302817
GordaRidge_Vent105_SUPRS9_2019          2.62814219 -2.9812951  3.0570484
GordaRidge_Vent041_SUPRS3_2019         -0.69592172 -4.2901449 -5.5715380
GordaRidge_Vent086_SUPRS1_2019          1.34187733 -4.2906166  0.2306734
GordaRidge_Vent107_SUPRS11_2019         4.10548120 -4.4581073  1.4173089
GordaRidge_Vent009_SUPRS1_2019          0.02649703  0.4100601  0.2067263
GordaRidge_Vent011_SUPRS3_2019         -2.10549955 -2.1273228  1.6182111
GordaRidge_Vent106_SUPRS10_2019         7.11184493  3.4292522 -1.7346496
GordaRidge_Vent010_SUPRS2_2019         -1.84599584 -0.2704446  1.1108578
GordaRidge_Vent087_SUPRS2_2019         -0.73350682  3.2922873 -5.7290464
                                              PC13        PC14       PC15
GordaRidge_Vent039_SUPRS1_2019          1.45574041 -1.36430736 -6.8306400
GordaRidge_Vent040_SUPRS2_2019         -1.81659395  2.01184113  6.3836722
GordaRidge_BSW081_sterivex_2019        -1.08036086  1.65489767  1.8269364
GordaRidge_BSW056_sterivex_2019_REPb    0.10883450  1.13767420 -0.6164449
GordaRidge_Vent088_SUPRS3_2019         -2.27047006  0.56597894 -1.1972686
GordaRidge_Plume001_sterivex_2019_REPa  0.53032296  0.22273083  0.7574762
GordaRidge_Plume001_sterivex_2019_REPb -0.01173338 -0.91437182 -0.5373933
GordaRidge_Plume036_sterivex_2019_REPb -1.34497343  0.93883523 -0.6440523
GordaRidge_Plume096_sterivex_2019       6.86147323  1.89215460  1.2990508
GordaRidge_Vent105_SUPRS9_2019         -0.90020743 -0.28145036  0.3948457
GordaRidge_Vent041_SUPRS3_2019          0.13314711  0.21061565  0.6181466
GordaRidge_Vent086_SUPRS1_2019          0.05806217  0.04563998 -0.0789418
GordaRidge_Vent107_SUPRS11_2019         4.34022224 -0.01293642  0.1991291
GordaRidge_Vent009_SUPRS1_2019          0.28173491 -2.24746909 -0.2760678
GordaRidge_Vent011_SUPRS3_2019         -0.78939081  4.53450744 -3.2682822
GordaRidge_Vent106_SUPRS10_2019        -6.77198632  0.68508248 -0.8179652
GordaRidge_Vent010_SUPRS2_2019         -0.72164897 -8.63300708  1.2616129
GordaRidge_Vent087_SUPRS2_2019          1.93782767 -0.44641602  1.5261862
                                             PC16        PC17          PC18
GordaRidge_Vent039_SUPRS1_2019          1.2832806  0.75016565  4.712845e-15
GordaRidge_Vent040_SUPRS2_2019         -0.2994072 -0.07895867  2.761361e-15
GordaRidge_BSW081_sterivex_2019         1.0173041  0.76670779 -1.355986e-14
GordaRidge_BSW056_sterivex_2019_REPb    0.7825043  0.60023506  2.493538e-15
GordaRidge_Vent088_SUPRS3_2019         -0.1435484 -0.12737550  2.780838e-15
GordaRidge_Plume001_sterivex_2019_REPa  4.3131427 -0.62686459 -2.221836e-15
GordaRidge_Plume001_sterivex_2019_REPb -7.3322085  1.01808193  2.729658e-15
GordaRidge_Plume036_sterivex_2019_REPb  0.5006400  0.65655982  2.626417e-15
GordaRidge_Plume096_sterivex_2019      -0.4083416  0.88511402  7.339894e-16
GordaRidge_Vent105_SUPRS9_2019          0.1197658  0.72729471 -6.217338e-15
GordaRidge_Vent041_SUPRS3_2019         -0.4930825 -0.15274411  4.177932e-15
GordaRidge_Vent086_SUPRS1_2019          0.1550844  0.33355042 -3.119402e-15
GordaRidge_Vent107_SUPRS11_2019         0.3250068  0.03251975  3.126441e-15
GordaRidge_Vent009_SUPRS1_2019         -0.8332094 -7.06865122  1.195868e-15
GordaRidge_Vent011_SUPRS3_2019         -0.4407791 -0.05969055  3.612404e-15
GordaRidge_Vent106_SUPRS10_2019         0.2927570  0.26516134 -1.216574e-14
GordaRidge_Vent010_SUPRS2_2019          0.8456788  2.02347829  4.365593e-15
GordaRidge_Vent087_SUPRS2_2019          0.3154124  0.05541588  5.188293e-15</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>pca_clr_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(clr_pca<span class="sc">$</span>x) <span class="sc">%&gt;%</span> </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"SAMPLE"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(SAMPLE)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># head(pca_clr_df)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Make PCA plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot PCA</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pca_clr_df) <span class="sc">+</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">fill =</span> VENT), <span class="at">size =</span> <span class="dv">4</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">'PC2 '</span>, <span class="fu">round</span>(lograt_variances[<span class="dv">2</span>,<span class="dv">2</span>]<span class="sc">*</span><span class="dv">100</span>,<span class="dv">2</span>),<span class="st">'%'</span>)) <span class="sc">+</span> <span class="co">#Extract y axis value from variance</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">'PC1 '</span>, <span class="fu">round</span>(lograt_variances[<span class="dv">1</span>,<span class="dv">2</span>]<span class="sc">*</span><span class="dv">100</span>,<span class="dv">2</span>),<span class="st">'%'</span>)) <span class="sc">+</span> <span class="co">#Extract x axis value from variance</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">'Dark2'</span>) <span class="sc">+</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">'Centered Log-Ratio PCA Ordination'</span>) <span class="sc">+</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="at">ratio =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in RColorBrewer::brewer.pal(n, pal): n too large, allowed maximum for palette Dark2 is 8
Returning the palette you asked for with that many colors</code></pre>
</div>
<div class="cell-output-display">
<p><img src="clustering_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="pcoa" class="level1">
<h1>PCoA</h1>
<p>PCoA is doing a PCA on a distance matrix constructed from the data. We then need a distance matrix. Different distance metrics emphasize separate attributes/factors in microbial community comparison.</p>
<section id="distance-matrix" class="level2">
<h2 class="anchored" data-anchor-id="distance-matrix">Distance matrix</h2>
<p>If we want to prioritize differences in presence/absence between samples - use Jaccard. This can be estimated from untransformed count data, and does a pretty good job considering rare taxa. Another is unweighted Unifrac that includes phylogenetic relatedness (see last week’s lesson).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: permute</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: lattice</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'lattice' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This is vegan 2.6-4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ape' was built under R version 4.2.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ape'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:compositions':

    balance</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    where</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>jac_dmat <span class="ot">&lt;-</span> <span class="fu">vegdist</span>(<span class="fu">t</span>(asv_df),<span class="at">method=</span><span class="st">"jaccard"</span>) <span class="co"># Jaccard dist metric</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>pcoa_jac <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">pcoa</span>(jac_dmat) <span class="co">#perform PCoA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can use the screeplot to determine if we are going in the right direction.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>jac_variances <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pcoa_jac<span class="sc">$</span>values<span class="sc">$</span>Relative_eig) <span class="sc">%&gt;%</span> </span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">PercVar =</span> <span class="st">'pcoa_jac.values.Relative_eig'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"PCaxis"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  data.frame</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(jac_variances, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.numeric</span>(PCaxis), <span class="at">y =</span> PercVar)) <span class="sc">+</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"grey"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>),</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"PC axis"</span>, <span class="at">y =</span> <span class="st">"% Variance"</span>, <span class="at">title =</span> <span class="st">"Log-Ratio PCoA Screeplot"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="clustering_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract variances from pcoa, from jaccard calculated dist. metric</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="do">## where samples fall among axes</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>pcoa_jac_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pcoa_jac<span class="sc">$</span>vectors) <span class="sc">%&gt;%</span> </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"TimeofDay"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(TimeofDay, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"excess"</span>, <span class="st">"TimeofDay"</span>), <span class="at">sep =</span> <span class="st">"_"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>excess) <span class="sc">%&gt;%</span> </span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>  data.frame</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Expected 2 pieces. Additional pieces discarded in 18 rows [1, 2, 3, 4, 5, 6, 7,
8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18].</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pcoa_jac_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  TimeofDay       Axis.1      Axis.2       Axis.3       Axis.4       Axis.5
1   Vent039 -0.511333960  0.24510098  0.002724799 -0.014780337  0.001248485
2   Vent040 -0.471925533  0.12993493  0.102824667  0.003851581 -0.018398321
3    BSW081  0.009154644  0.09012354 -0.526259280 -0.307777122  0.315467056
4    BSW056  0.063135741 -0.28182252 -0.035892201 -0.032751972 -0.056781429
5   Vent088  0.249164360  0.11271888  0.148602132  0.061580537  0.127690807
6  Plume001  0.273580285  0.31896957  0.052082558 -0.173870964 -0.235917650
       Axis.6       Axis.7        Axis.8      Axis.9      Axis.10     Axis.11
1  0.00575785 -0.002950044 -0.1035403386 -0.02064122 -0.043766331 -0.03282459
2 -0.02912210  0.031274317 -0.0408526685 -0.01139825  0.007774317  0.01496053
3 -0.07343554 -0.161919327  0.0008682262  0.02164156  0.037738741  0.04190189
4 -0.02412939  0.075845547 -0.1979221811  0.09264871  0.261368972 -0.34640665
5  0.04972538 -0.007616491 -0.1670897130  0.05159573 -0.074337976  0.11709158
6 -0.08461598  0.012419863  0.0722941976 -0.11208131 -0.061826255 -0.08292078
       Axis.12     Axis.13     Axis.14     Axis.15      Axis.16      Axis.17
1 -0.028074522 -0.03010100  0.17510883  0.21945956  0.109465415  0.075283311
2 -0.004368510  0.05237571 -0.05666681 -0.03448782 -0.124018240 -0.258128825
3  0.006443506  0.01841906 -0.03519646 -0.02368154 -0.010810246 -0.005225330
4  0.068221228  0.06867835 -0.02625598  0.04011823 -0.005187598  0.003907832
5  0.406026310 -0.01269803  0.04131232  0.01937920 -0.020588905 -0.007539995
6  0.005454713 -0.16190160 -0.22238807  0.12523584  0.015402771 -0.009846016</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Select eigen values from dataframe, round to 4 places. These will be the axes for the 3-D plot</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract variances from previously generated dataframe, round and multiply by 100 for plotting</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>eigenvalues<span class="ot">&lt;-</span><span class="fu">round</span>(jac_variances[,<span class="dv">2</span>], <span class="at">digits =</span> <span class="dv">4</span>)<span class="sc">*</span><span class="dv">100</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plotly - 3-D</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co"># plotly::plot_ly(pcoa_jac_df, type='scatter3d', mode='markers',</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co">#         x=~Axis.2,y=~Axis.3,z=~Axis.1,colors=~brewer.pal(6,'Set1'),color=~TimeofDay)%&gt;%</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a><span class="co">#   layout(font=list(size=18),</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a><span class="co">#          title='PCoA Jaccard Distance',</span></span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a><span class="co">#          scene=list(xaxis=list(title=paste0('Co 2 ',eigenvalues[2],'%'),</span></span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                                showticklabels=FALSE,zerolinecolor='black'),</span></span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a><span class="co">#                     yaxis=list(title=paste0('Co 3 ',eigenvalues[3],'%'),</span></span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a><span class="co">#                                showticklabels=FALSE,zerolinecolor='black'),</span></span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a><span class="co">#                     zaxis=list(title=paste0('Co 1 ',eigenvalues[1],'%'),</span></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a><span class="co">#                                showticklabels=FALSE,zerolinecolor='black')))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="euclidean" class="level2">
<h2 class="anchored" data-anchor-id="euclidean">Euclidean</h2>
<p>Performing the log-ratio transformation makes the data all occupy a similar dynamic range, so we can use magnitude-sensitive distances like Euclidean distance.</p>
<p><em>Euclidean</em>, recommended for analysis of the difference in compositions. e.g., when we are working to understand changes in relative abundance. Because it depends on the composition, we must input transformed data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ?dist()</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>euc_dmat <span class="ot">&lt;-</span> <span class="fu">dist</span>(asv_df, <span class="at">method =</span> <span class="st">"euclidean"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Conduct ordination w/distance matrix = PCoA</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>pcoa_euc<span class="ot">&lt;-</span>ape<span class="sc">::</span><span class="fu">pcoa</span>(euc_dmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Repeat screeplot check. Extract variances from pcoa, from jaccard calculated dist. metric</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>euc_variances <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pcoa_euc<span class="sc">$</span>values<span class="sc">$</span>Relative_eig) <span class="sc">%&gt;%</span> </span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">PercVar =</span> <span class="st">'pcoa_euc.values.Relative_eig'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"PCaxis"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  data.frame</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="co"># head(euc_variances)</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Screeplot check</span></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(euc_variances, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.numeric</span>(PCaxis), <span class="at">y =</span> PercVar)) <span class="sc">+</span> </span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"grey"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">size =</span> <span class="dv">10</span>),</span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>),</span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"PC axis"</span>, <span class="at">y =</span> <span class="st">"% Variance"</span>, <span class="at">title =</span> <span class="st">"Euclidean</span><span class="sc">\n</span><span class="st">Log-Ratio PCoA Screeplot"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="clustering_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<blockquote class="blockquote">
<p>Compare what the previous scree plot looked like.</p>
</blockquote>
<p>Is there a difference in the ordinate output when we use a different metric?</p>
<p>To compare with the euclidean distance ordination representing differences in relative composition. Further note: If your ordination has data which align in a ‘T’ or ‘+’ shape perpendicular to the axes this is often diagnostic of covariance attributed to the higher dimensions which are not plotted.</p>
<p><em>What happens if I have a screeplot that requires 3D, but I can’t do it?</em></p>
<p>Although our statistical ordinations appear to require at least 3 dimensions to communicate the data. However, we don’t always have the budget for a 3D plot. So we may want to impose the condition on an ordination technique that the answer MUST go in 2D. We turn to NMDS here.</p>
</section>
<section id="nmds" class="level2">
<h2 class="anchored" data-anchor-id="nmds">NMDS</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1984</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>So we can compare the relative composition based distance metric to the presence/absence based distance metric</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ?metaMDS()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>NMDS</em>: force data into a desired number of dimensions, works to preserve all pairwise distances between points.</p>
<p>Start with Euclidean distance again. But then transform with NMDS.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>euc_dmat <span class="ot">&lt;-</span> compositions<span class="sc">::</span><span class="fu">dist</span>(asv_df, <span class="at">method =</span> <span class="st">"euclidean"</span>) <span class="co"># From above</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>euc_nmds <span class="ot">&lt;-</span> vegan<span class="sc">::</span> <span class="fu">metaMDS</span>(euc_dmat, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">autotransform =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Run 0 stress 0.03176447 
Run 1 stress 0.0317942 
... Procrustes: rmse 0.01591051  max resid 0.2519757 
Run 2 stress 0.03271511 
Run 3 stress 0.03191601 
... Procrustes: rmse 0.01869626  max resid 0.3884156 
Run 4 stress 0.03261754 
Run 5 stress 0.03334667 
Run 6 stress 0.0355446 
Run 7 stress 0.0332108 
Run 8 stress 0.03165351 
... New best solution
... Procrustes: rmse 0.0197889  max resid 0.4094653 
Run 9 stress 0.0307389 
... New best solution
... Procrustes: rmse 0.01601655  max resid 0.3265687 
Run 10 stress 0.03206538 
Run 11 stress 0.03114016 
... Procrustes: rmse 0.01832106  max resid 0.3892465 
Run 12 stress 0.03297826 
Run 13 stress 0.02989012 
... New best solution
... Procrustes: rmse 0.02733152  max resid 0.548664 
Run 14 stress 0.03282565 
Run 15 stress 0.03122517 
Run 16 stress 0.03259375 
Run 17 stress 0.03179146 
Run 18 stress 0.03220929 
Run 19 stress 0.03184396 
Run 20 stress 0.03285802 
*** Best solution was not repeated -- monoMDS stopping criteria:
    20: no. of iterations &gt;= maxit</code></pre>
</div>
</div>
<p>Repeat with the Jaccard transformation</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>jac_dmat <span class="ot">&lt;-</span> vegan<span class="sc">::</span> <span class="fu">vegdist</span>(<span class="fu">t</span>(asv_df), <span class="at">method =</span> <span class="st">"jaccard"</span>) <span class="co"># From above</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>jac_nmds<span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">metaMDS</span>(jac_dmat, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">autotransform =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Run 0 stress 0.1423445 
Run 1 stress 0.1919306 
Run 2 stress 0.1373766 
... New best solution
... Procrustes: rmse 0.04095258  max resid 0.1313601 
Run 3 stress 0.1811173 
Run 4 stress 0.1375691 
... Procrustes: rmse 0.01209244  max resid 0.03782542 
Run 5 stress 0.1828614 
Run 6 stress 0.1939918 
Run 7 stress 0.1933375 
Run 8 stress 0.2179407 
Run 9 stress 0.1734027 
Run 10 stress 0.1811173 
Run 11 stress 0.1607189 
Run 12 stress 0.1373766 
... New best solution
... Procrustes: rmse 1.788908e-05  max resid 4.848831e-05 
... Similar to previous best
Run 13 stress 0.1423445 
Run 14 stress 0.1610092 
Run 15 stress 0.1375691 
... Procrustes: rmse 0.01208686  max resid 0.03781589 
Run 16 stress 0.1827459 
Run 17 stress 0.180109 
Run 18 stress 0.1676982 
Run 19 stress 0.1424365 
Run 20 stress 0.2052544 
*** Best solution repeated 1 times</code></pre>
</div>
</div>
<p>Take a look at stress - overall this value is not extremely informative, but know that the closer stress is to 1 the less representative of your actual data the NMDS is.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>euc_nmds<span class="sc">$</span>stress </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.02989012</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>jac_nmds<span class="sc">$</span>stress</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1373766</code></pre>
</div>
</div>
<p>Additionally, the axes for NMDS are totally arbitrary, so axis scaling does not matter and data can be rotated/reflected about axes and the NMDS is still the same euc_nmds$points #Extract points from NMDS</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>euc_frame <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(euc_nmds<span class="sc">$</span>points) <span class="sc">%&gt;%</span> </span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"TimeofDay"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(TimeofDay, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"excess"</span>, <span class="st">"TimeofDay"</span>), <span class="at">sep =</span> <span class="st">"_"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>excess) <span class="sc">%&gt;%</span> </span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  data.frame</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Expected 2 pieces. Missing pieces filled with `NA` in 1000 rows [1, 2, 3, 4, 5,
6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, ...].</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Plotting euclidean distance NMDS</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(euc_frame,<span class="fu">aes</span>(<span class="at">x =</span> MDS1, <span class="at">y =</span> MDS2, <span class="at">fill =</span> TimeofDay)) <span class="sc">+</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">4</span>, <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="fu">aes</span>(<span class="at">fill =</span> TimeofDay)) <span class="sc">+</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Dark2"</span>) <span class="sc">+</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"NMDS 1"</span>, <span class="at">y =</span> <span class="st">"NMDS 2"</span>, <span class="at">title =</span> <span class="st">"Euclidean Distance NMDS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="clustering_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./customize-plots.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Customize plots</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./correlation_linearregression.html" class="pagination-link">
        <span class="nav-page-text">Correlation &amp; Linear regression</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>